romanisim/tests/test_l3.py

```diff
@@ -0,0 +1,133 @@
+from astropy.coordinates import SkyCoord
+from astropy import units as u
+from astropy.time import Time
+from astropy import table
+import asdf
+import webbpsf
+from astropy.modeling.functional_models import Sersic2D
+import pytest
+from metrics_logger.decorators import metrics_logger
+from romanisim import log
+from roman_datamodels.stnode import WfiScienceRaw, WfiImage
+import roman_datamodels.maker_utils as maker_utils
+import romanisim.bandpass
+
+
+@metrics_logger("DMS232")
+@pytest.mark.soctests
+def test_inject_sources_into_mosaic():
+    """Inject sources into a mosaic.
+    """
+
+    # Set constants and metadata
+    galsim.roman.n_pix = 200
+    rng_seed = 42
+    metadata = copy.deepcopy(parameters.default_mosaic_parameters_dictionary)
+    metadata['basic']['optical_element'] = 'F158'
+
+    # Create WCS
+    twcs = wcs.get_mosaic_wcs(metadata, shape=(galsim.roman.n_pix, galsim.roman.n_pix))
+
+    # Create initial Level 3 mosaic
+
+    # Create Four-quadrant pattern of gaussian noise, centered around one
+    # Each quadrant's gaussian noise scales like total exposure time
+    # (total files contributed to each quadrant)
+
+    # Create gaussian noise generators
+    g1 = galsim.GaussianDeviate(rng_seed, mean=1.0, sigma=0.01)
+    g2 = galsim.GaussianDeviate(rng_seed, mean=1.0, sigma=0.02)
+    g3 = galsim.GaussianDeviate(rng_seed, mean=1.0, sigma=0.05)
+    g4 = galsim.GaussianDeviate(rng_seed, mean=1.0, sigma=0.1)
+
+    # Create level 3 mosaic model
+    l3_mos = maker_utils.mk_level3_mosaic(shape=(galsim.roman.n_pix, galsim.roman.n_pix))
+
+    # Update metadata in the l3 model
+    for key in metadata.keys():
+        if key in l3_mos.meta:
+            l3_mos.meta[key].update(metadata[key])
+
+    # Populate the mosaic data array with gaussian noise from generators
+    g1.generate(l3_mos.data.value[0:100, 0:100])
+    g2.generate(l3_mos.data.value[0:100, 100:200])
+    g3.generate(l3_mos.data.value[100:200, 0:100])
+    g4.generate(l3_mos.data.value[100:200, 100:200])
+
+    # Define Poisson Noise of mosaic
+    l3_mos.var_poisson.value[0:100, 0:100] = 0.01**2
+    l3_mos.var_poisson.value[0:100, 100:200] = 0.02**2
+    l3_mos.var_poisson.value[100:200, 0:100] = 0.05**2
+    l3_mos.var_poisson.value[100:200, 100:200] = 0.1**2
+
+    # Create normalized psf source catalog (same source in each quadrant)
+    sc_dict = {"ra": 4 * [0.0], "dec": 4 * [0.0], "type": 4 * ["PSF"], "n": 4 * [-1.0],
+               "half_light_radius": 4 * [0.0], "pa": 4 * [0.0], "ba": 4 * [1.0], "F158": 4 * [1.0]}
+    sc_table = table.Table(sc_dict)
+
+    # Set locations
+    xpos_idx = [50, 50, 150, 150]
+    ypos_idx = [50, 150, 50, 150]
+
+    # Populate flux scaling ratio and catalog
+    Ct = []
+    for idx, (x, y) in enumerate(zip(xpos_idx, ypos_idx)):
+        # Set scaling factor for injected sources
+        # Flux / sigma_p^2
+        if l3_mos.var_poisson[x][y].value != 0:
```

Comment:
I think I at least ended up thinking we needed [y, x] here.

Reply:
Corrected.